flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/PipelinedSubpartition.java

```diff
@@ -302,7 +307,7 @@ public int unsynchronizedGetNumberOfQueuedBuffers() {
 	public void flush() {
 		final boolean notifyDataAvailable;
 		synchronized (buffers) {
-			if (buffers.isEmpty()) {
+			if (buffers.isEmpty() || readView == null || readView.isBlockedByCheckpoint()) {
```

Comment:
I know it can avoid entering the following actions if `readerView` is not fitted.  But another issue is that `notifyDataAvailable` would be called in three places in this class, and it already judges the `readerView` inside for unification. So I suggested to add the condition `readView.isBlockedByCheckpoint()` inside `notifyDataAvailable`.

Reply:
Some tags should be  checked in lock, I removed the check in notifyDataAvailable.